### PR TITLE
Identify when a stmt could have been parsed as an expr

### LINKED ISSUE #47287: Returning a comparison expression with a match fails depending on the operand order
Suppose the return expression of a function is a comparison between a match expression and some other operand. If the operand and comparison operator come before the match expression, the code compiles. However, if the operand and comparison operator come _after_ the match expression, the code fails to compile.

```rust
fn works(x: u32) -> bool {
    0 < match x {
        _ => 1,
    }
}

fn fails(x: u32) -> bool {
    match x {
        _ => 1,
    } > 0
}
```
The error is thus:
```
error: expected expression, found `>`
  --> src/main.rs:10:7
   |
10 |     } > 0
   |       ^

error[E0308]: match arms have incompatible types
  --> src/main.rs:8:5
   |
8  | /     match x {
9  | |         _ => 1,
10 | |     } > 0
   | |_____^ expected bool, found integral variable
   |
   = note: expected type `bool`
              found type `{integer}`
note: match arm with an incompatible type
  --> src/main.rs:9:14
   |
9  |         _ => 1,
   |              ^
```
Strangely, the code does work if the comparison is a statement:
```rust
fn wontfail(x: u32) -> bool {
    let y = match x {
        _ => 1,
    } > 0;
    y
}
```
Tested using rustc `1.23.0` on Playground.

### LINKED ISSUE #54186: Two consecutive if lets cannot be used as function result
[MCVE](https://play.rust-lang.org/?gist=7a007ac374bd52313cfeaa58ef6c09e2&version=stable&mode=debug&edition=2015)

    fn nowork(a: Option<u32>, b: Option<u32>) -> bool {
        if let Some(x) = a { true } else { false }
        &&
        if let Some(y) = a { true } else { false }
    }
    
    fn work(a: Option<u32>, b: Option<u32>) -> bool {
        let z = if let Some(x) = a { true } else { false }
        &&
        if let Some(y) = a { true } else { false };
        z
    }

Error:

    error[E0308]: mismatched types
     --> src/main.rs:2:26
      |
    2 |     if let Some(x) = a { true } else { false }
      |                          ^^^^ expected (), found bool
      |
      = note: expected type `()`
                 found type `bool`

    error[E0308]: mismatched types
     --> src/main.rs:2:40
      |
    2 |     if let Some(x) = a { true } else { false }
      |                                        ^^^^^ expected (), found bool
      |
      = note: expected type `()`
                 found type `bool`

    error[E0308]: mismatched types
     --> src/main.rs:3:5
      |
    1 |   fn nowork(a: Option<u32>, b: Option<u32>) -> bool {
      |                                                ---- expected `bool` because of return type
    2 |       if let Some(x) = a { true } else { false }
    3 | /     &&
    4 | |     if let Some(y) = a { true } else { false }
      | |______________________________________________^ expected bool, found &&bool
      |
      = note: expected type `bool`
                 found type `&&bool`

Am I missing something? Why is the first one not allowed, but the second one?

### LINKED ISSUE #54482: Failure to parse `{2} + {2}`
These examples should compile according to the rust reference:

```rust
fn compiles() -> i32 {
    ({2}) + ({2})
}

fn fails() -> i32 {
    {2} + {2}
}

fn compiles2() -> i32 {
    2 + {2}
}

fn fails2() -> i32 {
    {2} + 2
}
```

([Playground](https://play.rust-lang.org/?gist=36a06367e5e9b2eded3b7a15067306d5&version=stable&mode=debug&edition=2015))

Errors:

```
   Compiling playground v0.0.1 (file:///playground)
error: expected expression, found `+`
 --> src/lib.rs:6:9
  |
6 |     {2} + {2}
  |         ^ expected expression

error: expected expression, found `+`
  --> src/lib.rs:14:9
   |
14 |     {2} + 2
   |         ^ expected expression

error: aborting due to 2 previous errors

error: Could not compile `playground`.

To learn more, run the command again with --verbose.

```

Rust reference snippet (thanks @Moongoodboy-K for the help here):
> Expression : BlockExpression | OperatorExpression | ..
> OperatorExpression : ArithmeticOrLogicalExpression | ..
> ArithmeticOrLogicalExpression : Expression `+` Expression | ..
> BlockExpression : `{` InnerAttribute* Statement* Expression? `}`

cc: @Moongoodboy-K

### LINKED ISSUE #59975: Redundant braces lead to compilation failure: expected expression, found `as`
The following program surprisingly does not compile:
```rust
fn main() {
    let x = &mut 2;
    let _val = {{x} as *mut i32};
}
```
The error is rather confusing, somehow it expects something to be `()`:
```
error: expected expression, found keyword `as`
 --> src/main.rs:3:21
  |
3 |     let _val = {{x} as *mut i32};
  |                     ^^ expected expression

error[E0308]: mismatched types
 --> src/main.rs:3:18
  |
1 | fn main() {
  |           - expected `()` because of default return type
2 |     let x = &mut 2;
3 |     let _val = {{x} as *mut i32};
  |                  ^ expected (), found mutable reference
  |
  = note: expected type `()`
             found type `&mut {integer}`
```
If I remove either of the pair of braces in the last line, it works as expected.

### PR DESCRIPTION
**Pull request by estebank**
There are some expressions that can be parsed as a statement without
a trailing semicolon depending on the context, which can lead to
confusing errors due to the same looking code being accepted in some
places and not others. Identify these cases and suggest enclosing in
parenthesis making the parse non-ambiguous without changing the
accepted grammar.

Fix #54186, cc #54482, fix #59975, fix #47287.

### CODE REVIEW

**Comment by varkor** [comment]
> Fix #54186, fix #54482, fix #59975.

I'm not sure it's fair to say it actually fixes these, though it improves the diagnostics. I think it's possible we want to accept the code there (there isn't a reason these are not accepted other than "it's hard to parse").

**Comment by estebank** [comment]
@varkor I thought that we wanted to make sure the parser was unambiguous. Some of these cases can be identified in the parser itself, but some require checks in the type checker or elsewhere. At that point the options are either only accept _some_ of these operations but not others, which could look arbitrary, or have a more advanced parser/evaluator, which aiui is out of the question.

**Comment by varkor** [comment]
Sorry for the delay; I will get to this soon.

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -4165,9 +4165,31 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                                let sp = tcx.sess.source_map().start_point(expr.span);
+                                if let Some(sp) = tcx.sess.parse_sess.abiguous_block_expr_parse
+                                    .borrow().get(&sp)
+                                {
+                                    if let Ok(snippet) = tcx.sess.source_map()
+                                        .span_to_snippet(*sp)
+                                    {
+                                        err.span_suggestion(
+                                            *sp,
+                                            "parenthesis are required to parse this \
```
Comment (varkor) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
                                            "parentheses are required to parse this \
```

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -2925,6 +2927,19 @@ impl<'a> Parser<'a> {
...
                                               self.this_token_descr());
                             let mut err = self.fatal(&msg);
+                            let sp = self.sess.source_map().start_point(self.span);
+                            if let Some(sp) = self.sess.abiguous_block_expr_parse.borrow()
+                                .get(&sp)
+                            {
+                                if let Ok(snippet) = self.sess.source_map().span_to_snippet(*sp) {
+                                    err.span_suggestion(
+                                        *sp,
+                                        "parenthesis are required to parse this as an expression",
```
Comment (varkor) [inline src/libsyntax/parse/parser.rs]:
```suggestion
                                        "parentheses are required to parse this as an expression",
```

File: src/libsyntax/parse/parser.rs [hunk 3]
```diff
@@ -3616,9 +3631,42 @@ impl<'a> Parser<'a> {
...
+                let mut err = self.sess.span_diagnostic.struct_span_err(self.span, &format!(
+                    "expected expression, found `{}`",
+                    pprust::token_to_string(&self.token),
+                ));
+                err.span_label(self.span, "expected expression");
+                let snippet = self.sess.source_map().span_to_snippet(lhs.span)
+                    .unwrap_or_else(|_| pprust::expr_to_string(&lhs));
+                err.span_suggestion(
+                    lhs.span,
+                    "parenthesis are required to parse this as an expression",
```
Comment (varkor) [inline src/libsyntax/parse/parser.rs]:
```suggestion
                    "parentheses are required to parse this as an expression",
```

File: src/libsyntax/parse/parser.rs [hunk 4]
```diff
@@ -4929,6 +4977,17 @@ impl<'a> Parser<'a> {
                         );
                         let mut err = self.fatal(&msg);
                         err.span_label(self.span, format!("expected {}", expected));
+                        let sp = self.sess.source_map().start_point(self.span);
+                        if let Some(sp) = self.sess.abiguous_block_expr_parse.borrow().get(&sp) {
+                            if let Ok(snippet) = self.sess.source_map().span_to_snippet(*sp) {
+                                err.span_suggestion(
+                                    *sp,
+                                    "parenthesis are required to parse this as an expression",
```
Comment (varkor) [inline src/libsyntax/parse/parser.rs]:
```suggestion
                                    "parentheses are required to parse this as an expression",
```

File: src/librustc_typeck/check/mod.rs [hunk 5]
```diff
@@ -4165,9 +4165,31 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                                let mut err = type_error_struct!(
+                                    tcx.sess,
+                                    expr.span,
+                                    oprnd_t,
+                                    E0614,
+                                    "type `{}` cannot be dereferenced",
+                                    oprnd_t,
+                                );
+                                let sp = tcx.sess.source_map().start_point(expr.span);
+                                if let Some(sp) = tcx.sess.parse_sess.abiguous_block_expr_parse
```
Comment (varkor) [inline src/librustc_typeck/check/mod.rs]: This pattern is repeated several times (the one emitting the `"parentheses are required to parse this as an expression"` message). Could it be pulled out?

File: src/libsyntax/parse/parser.rs [hunk 6]
```diff
@@ -3616,9 +3631,42 @@ impl<'a> Parser<'a> {
...
-        if self.expr_is_complete(&lhs) {
-            // Semi-statement forms are odd. See https://github.com/rust-lang/rust/issues/29071
-            return Ok(lhs);
+        match (self.expr_is_complete(&lhs), AssocOp::from_token(&self.token)) {
+            (true, None) => {
+                // Semi-statement forms are odd. See https://github.com/rust-lang/rust/issues/29071
+                return Ok(lhs);
+            }
+            (false, _) => {} // continue parsing the expression
+            (true, Some(AssocOp::Multiply)) | // `{ 42 } *foo = bar;`
```
Comment (varkor) [inline src/libsyntax/parse/parser.rs]: Presumably these aren't exhaustive? It seems a little ad hoc to match on a specific set of operators like this. Is there a more general way we could detect this?

**Review by varkor — COMMENTED** [review]
I made a few comments, but I still need to think about this some more.

**Comment by varkor** [comment]
> @varkor I thought that we wanted to make sure the parser was unambiguous. Some of these cases can be identified in the parser itself, but some require checks in the type checker or elsewhere.

I agree that the grammar should be unambiguous, but I was under the impression some of these cases were not ambiguous; they were just tricky to detect. Maybe it's reasonable to close these issues as diagnostic issues, but I think there should be an issue somewhere keeping track of this until a deliberate decision has been made.

**Comment by estebank** [comment]
> I agree that the grammar should be unambiguous, but I was under the impression some of these cases were not ambiguous; they were just tricky to detect.

That is indeed true. If you look at https://github.com/rust-lang/rust/pull/60188/files#diff-11be04506fe70f5f173b5fdf7f294ad7 you can see the full list of unambiguous operations, but I wouldn't want to have some constructs work, and some not in an _almost_ arbitrary manner.

File: src/libsyntax/parse/parser.rs (hunk 6 above)
Comment (estebank) [reply src/libsyntax/parse/parser.rs]: The exhaustive list is checked in [line 3648](https://github.com/rust-lang/rust/pull/60188/files#diff-da9d34ca1d0f6beee2838cf02e07345cR3648). These are checked first because they are ambiguous but also very common source of trouble, so we want to keep their info around in `abiguous_block_expr_parse` to improve diagnostics in the appropriate case in later stages. I can move these three to another method like `can_continue_expr_unambiguously` if you wish.

File: src/libsyntax/parse/lexer/mod.rs [hunk 7]
```diff
@@ -1918,6 +1918,7 @@ mod tests {
             raw_identifier_spans: Lock::new(Vec::new()),
             registered_diagnostics: Lock::new(ErrorMap::new()),
             buffered_lints: Lock::new(vec![]),
+            abiguous_block_expr_parse: Lock::new(FxHashMap::default()),
```
Comment (varkor) [inline src/libsyntax/parse/lexer/mod.rs]:
```suggestion
            ambiguous_block_expr_parse: Lock::new(FxHashMap::default()),
```

File: src/libsyntax/parse/parser.rs (hunk 6 above)
Comment (varkor) [reply src/libsyntax/parse/parser.rs]: Maybe just a comment explaining this above?

File: src/libsyntax/util/parser.rs [hunk 8]
```diff
@@ -207,6 +207,28 @@ impl AssocOp {
             ObsoleteInPlace | Assign | AssignOp(_) | As | DotDot | DotDotEq | Colon => None
         }
     }
+
+    pub fn can_continue_expr_unambiguously(&self) -> bool {
```
Comment (varkor) [inline src/libsyntax/util/parser.rs]: It'd be good to have a comment explaining what this function does here.

File: src/libsyntax/parse/parser.rs (hunk 6 above)
Comment (estebank) [reply src/libsyntax/parse/parser.rs]: done

File: src/libsyntax/parse/lexer/mod.rs (hunk 7 above)
Comment (estebank) [reply src/libsyntax/parse/lexer/mod.rs]: done

File: src/libsyntax/util/parser.rs (hunk 8 above)
Comment (estebank) [reply src/libsyntax/util/parser.rs]: done

**Comment by estebank** [comment]
ping @varkor :)
